### PR TITLE
fixes #8613 - support upgrades in the installer 

### PR DESCRIPTION
**Pull request by stbenjam**

### CODE REVIEW

**Comment by ehelms** [comment]
Chatted briefly in person - I am ok with the straight `--upgrade` option to upgrade Katello from any point to latest released version.

**Comment by stbenjam** [comment]
@ehelms Needs another review if you can.

Can this make 2.1 or do I need to ajdust https://github.com/Katello/katello.org/pull/90? I think it's worthwhile to not have users do all the steps manually.

I still need to test a capsule upgrade, but this should work for it too.

**Comment by stbenjam** [comment]
Successfully upgraded both a Katello and a Capsule using this PR and the instructions at https://github.com/Katello/katello.org/pull/95

**Comment by stbenjam** [comment]
[test]

**Comment by ehelms** [comment]
@stbenjam spinning up environments, will test this either today or tomorrow morning.

File: hooks/pre/30-upgrade.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+end
+
+def upgrade_step(step)
+  noop = app_value(:noop) ? ' (noop)' : ''
+  Kafo::KafoConfigure.logger.info "Upgrade Step: #{step.to_s}" + noop
+  send(step) unless app_value(:noop)
+end
+
+if app_value(:upgrade)
+  say "<%= color('Upgrading...', :info) %>"
```
Comment (ehelms) [inline hooks/pre/30-upgrade.rb]: Maybe something more like `Katello upgrading...` or `The application is upgrading...` or `Starting upgrade procedure for application/Katello...`. Perhaps include some output per step? Mine is just showing `Upgrading...` which, as long as it has been proceeding, and without any progress output, I am of the mind it is stuck.

**Comment by ehelms** [comment]
My el6 output:

```
Upgrading...
Installing             --- /var/lib/pgsql/data/pg_hba.confrver::2015-01-12 20: [65%] [....................................................................................                                          /Stage[main]/Foreman::Database/Foreman::Rake[db:migrate]/Exec[foreman-rake-db:migrate]: Failed to call refresh: /usr/sbin/foreman-rake db:migrate returned 1 instead of one of [0]
 /Stage[main]/Foreman::Database/Foreman::Rake[db:migrate]/Exec[foreman-rake-db:migrate]: /usr/sbin/foreman-rake db:migrate returned 1 instead of one of [0]
 /Stage[main]/Foreman::Database/Foreman::Rake[db:seed]/Exec[foreman-rake-db:seed]: Failed to call refresh: /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
 /Stage[main]/Foreman::Database/Foreman::Rake[db:seed]/Exec[foreman-rake-db:seed]: /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
 /Stage[main]/Foreman::Database/Foreman::Rake[apipie:cache]/Exec[foreman-rake-apipie:cache]: Failed to call refresh: /usr/sbin/foreman-rake apipie:cache returned 1 instead of one of [0]
 /Stage[main]/Foreman::Database/Foreman::Rake[apipie:cache]/Exec[foreman-rake-apipie:cache]: /usr/sbin/foreman-rake apipie:cache returned 1 instead of one of [0]
 /usr/share/foreman/script/foreman-config      -k ssl_ca_file -v '/etc/foreman/proxy_ca.pem'      -k ssl_certificate -v '/etc/foreman/client_cert.pem'      -k ssl_priv_key -v '/etc/foreman/client_key.pem' returned 1 instead of one of [0]
 /Stage[main]/Certs::Foreman/Exec[foreman_certs_config]/returns: change from notrun to 0 failed: /usr/share/foreman/script/foreman-config      -k ssl_ca_file -v '/etc/foreman/proxy_ca.pem'      -k ssl_certificate -v '/etc/foreman/client_cert.pem'      -k ssl_priv_key -v '/etc/foreman/client_key.pem' returned 1 instead of one of [0]
 /Stage[main]/Foreman_proxy::Register/Foreman_smartproxy[katello-centos6-2.0.example.com]: Could not evaluate: Could not load data from https://katello-centos6-2.0.example.com
Installing             Done                                               [100%] [...............................................................................................................................]
  Something went wrong! Check the log for ERROR-level output
  The full log is at /var/log/katello-installer/katello-installer.log
Performing post-upgrade tasks...
/usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:104:in `error': wrong number of arguments (6 for 1) (ArgumentError)
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:104:in `send'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:104:in `dump_buffer'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:104:in `each'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:104:in `dump_buffer'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:103:in `each'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:103:in `dump_buffer'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:94:in `dump_errors'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/logger.rb:99:in `dump_errors'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/exit_handler.rb:26:in `exit'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/kafo_configure.rb:122:in `exit'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/kafo_configure.rb:338:in `run_installation'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/kafo_configure.rb:109:in `execute'
	from /usr/lib/ruby/gems/1.8/gems/clamp-0.6.2/lib/clamp/command.rb:68:in `run'
	from /usr/lib/ruby/gems/1.8/gems/clamp-0.6.2/lib/clamp/command.rb:126:in `run'
	from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.5/lib/kafo/kafo_configure.rb:116:in `run'
	from /usr/sbin/katello-installer:9
```

**Comment by stbenjam** [comment]
@ehelms I haven't seen that on any of my el6 upgrades, can I see katello-installer.log and foreman-debug? Will take a look tomorrow.

**Comment by stbenjam** [comment]
I can't reproduce this on el6, and for some reason the debug log you sent me doesn't have a complete installer logger.  I'm really interested to figure out why yours failed.

```
[root@katello-centos6-bats katello-installer]# bin/katello-installer --upgrade
Upgrading...
Installing             Done                                               [100%] [..........................................................................................................................................................]
  Success!
  * Katello is running at https://katello-centos6-bats.example.com
      Initial credentials are admin / iA38EBdofBFGz4HR
  * Capsule is running at https://katello-centos6-bats.example.com:9090
  * To install additional capsule on separate machine continue by running:"

      capsule-certs-generate --capsule-fqdn "$CAPSULE" --certs-tar "~/$CAPSULE-certs.tar"

  The full log is at /var/log/katello-installer/katello-installer.log
Performing post-upgrade tasks...
Katello upgrade completed!
```

File: hooks/pre/30-upgrade.rb (hunk 1 above)
Comment (stbenjam) [inline hooks/pre/30-upgrade.rb]: Ok, made more verbose for the user to know what's going on.

**Comment by stbenjam** [comment]
Also, when doing the upgrade, you'll need to copy `/etc/katello-installer/answers.katello-installer.yaml` to the config dir of the checkout, and it seems to work pretty reliably for me.

After wrangling a 2.0 install on CentOS 7, I also got a clean upgrade to nightly.  I haven't tried to upgrade to 2.1 RC1 because the installer was broken , but may work now if @jlsherrill pushed the new installer out.

**Comment by ehelms** [comment]
Everything seems fine, except for the DockerImage error, which I can't entirely tell if that is related to the differences between the nightly installer and the 2.1 branch. I can try cherry-picking this on top of that branch and doing scratch builds instead of relying on the scrach builds here.

**Comment by stbenjam** [comment]
I've successfully upgraded a capsule now as well.  That requires https://github.com/Katello/katello-installer/pull/166

@ehelms Please review again

**Comment by ehelms** [comment]
@stbenjam You can now use foreman-config to set that there are pending db migrations and seed such that the installer will re-run those steps. What are your thoughts on making use of that vs running how you have them currently?

**Comment by stbenjam** [comment]
Oh, way to throw a :wrench: in the works!  I think it's a good idea.. but we tested the PR a bunch as is, and I really wanted to get a proof of concept up of qpid dispatch-router...

Is it just running foreman-config in the upgrade pre-hook to tell Foreman to do the seed/migration?

**Comment by ehelms** [comment]
@stbenjam yessir, and I considered that we tested. Given this is setting our base for future upgrades, I just thought I'd ask. We could accept this and then try a smaller diff with the other way of doing things that way we have something that works and can switch over if the config way works? (Running a test of this PR as we speak FYI)

**Comment by stbenjam** [comment]
Tried but:
```
foreman-config script is deprecated. Please consider using `foreman-rake config` instead
```

It also looks like puppet-foreman hasn't even moved to using `foreman-rake config` yet -- maybe we should wait until they get it sorted?  

I'd like to revisit the topic of doing it in puppet as well, like you originally wanted to do - then we can simplify the hooks if we can get pulp and candlepin migrations done.

**Comment by ehelms** [comment]
Thanks for looking into it - ACK!

**Comment by stbenjam** [comment]
Yay! :tada:

**Pull request merged**